### ISSUE TITLE
Feature request: provide support for PMTiles format files?

### ISSUE THREAD

**Issue by anders86755** [onthegomap/maplibre-contour#212]
About PMTiles format : https://github.com/protomaps/PMTiles

The current initialization of demSource is as follows:

```
const demSource = new mlcontour.DemSource({
        url: 'https://demotiles.maplibre.org/terrain-tiles/{z}/{x}/{y}.png',
        encoding: 'mapbox',
        maxzoom: 12,
        worker: true
    });
```
If support for pmtiles format could be implemented, for example:

```
const demSource = new mlcontour.DemSource({
        url: 'pmtiles://https://xxx.xxx.com/xxx.pmtiles',
        encoding: 'mapbox',
        maxzoom: 12,
        worker: true
    });
```
The pmtiles format is very promising. It would be great if your plug-in could support this format.

Anyway，Thanks for this great plugin :)

**Comment by Patukas** [comment]
Same case here.

I am using pmtiles for dem source.

I am interested in use maplibre-contour.

Thanks.

**Comment by msbarry** [comment]
Yep! This is a known limitation of the plugin. I'm working to move this functionality into maplibre so you don't need this plugin anymore: https://github.com/maplibre/maplibre-style-spec/issues/583. Then you'll be able to use the pmtiles plugin for dem tiles that power a built-in contour source you can render line and label layers from.

**Comment by HarelM** [comment]
While this may take longer than expected, maybe it will be simpler to allow overriding the `fetch` method in non-worker case as a temporary solution? or add the ability to load a script to the worker in worker mode?
Let me know if I can help out here in any way as I'm also interested in something similar to this feature request.

**Comment by acalcutt** [comment]
I spent some time yesterday to see if I could add in the pmtiles npm plugin into maplibre-contour and modify its fetchTile to use pmtiles.getZxy(). I am testing with [this branch](https://github.com/acalcutt/maplibre-contour/tree/pmtiles_test) and you can see the changes I have made so far at https://github.com/onthegomap/maplibre-contour/compare/main...acalcutt:maplibre-contour:pmtiles_test 

I think I have it at least reading the tiles, because when testing on https://wifidb.net/demo/pmtiles/contour/ I can see it is loading the hillshade , which is coming from the `tiles: [demSource.sharedDemProtocolUrl]` , which is reading the pmtiles file through maplibre-countour (there is no separate pmtiles plugin on that page).

However right now I haven't been able to get the contour part of this working. I can see from the console.log's I added that it is trying to load them. it is getting neighbor tiles, but the isolines seem to be blank at the end. I am wondering if maybe this relates to not specifying a tile type when i convert the pmtiles array buffer to blob....but I am not really sure.
![image](https://github.com/user-attachments/assets/5aa8a57b-b735-4785-bb06-09828e896cb0)

I am honestly not the best with typscript or promises so I have been asking google ai a lot of questions just to get this far, so if anyone has any ideas to get get this working I am up for suggestions. The PMtiles support is somewhat based on the work i did in tileserver-gl at https://github.com/maptiler/tileserver-gl/blob/master/src/pmtiles_adapter.js

**Comment by acalcutt** [comment]
Actually it looks like I did at least get it working. It was failing because my source was 'terrarium' but I had it set to 'mapbox' in the DemSource

https://wifidb.net/demo/pmtiles/contour/
![image](https://github.com/user-attachments/assets/e6043f62-05a2-449b-90f7-08af00350598)

**Comment by gvkhna** [comment]
How are you generating the pmtiles, is it based on the open contour dataset?

**Comment by acalcutt** [comment]
One nice thing about maplibre-cuntour is that even though the dem only goes to zoom 8, you are still able to use higer zoom levels on your lines and still get nice crisp looking lines at high zoom levels

**Comment by gvkhna** [comment]
Unrelated but I wanted to ask what specific advantages of PMTiles are you all looking for, for this purpose. Is it simplicity of one file. I'm encountering the issues related to the browser is not caching range requests at all, but it does cache the png tiles fine which helps reduce network traffic. So i'm just trying to understand what benefits the PMTiles format would provide in this use case, besides perhaps performance?

**Comment by HarelM** [comment]
I've created the following PR: #359 to allow both replacing the fetch tile and decode image method for local dem manager, which will allow running this plugin in node environment for example.

**Comment by acalcutt** [comment]
For anyone interested in pmtiles support with maplibre-contour, I have made 2 projects which can give you a starting point

The first is [Contour-Generator](https://github.com/acalcutt/contour_generator/), which uses the feature added in https://github.com/onthegomap/maplibre-contour/pull/359 to allow for local pmtiles access to pmtiles in node. It just uses the maplibre-contour npm package and sets its own [decodeImage and getTile functions].  (https://github.com/acalcutt/contour_generator/blob/main/src/generate-countour-tile-pyramid.ts#L168-L193)

The seconds is [maplibre-contour-pmtiles](https://github.com/acalcutt/maplibre-contour-pmtiles) which is a continuation of the code I showed above, which integrates pmtiles plugin directly into the maplibre-contour. I had made this so it workied with node with local pmtiles, but when built with rollup there is some code to [switch to library](https://github.com/acalcutt/maplibre-contour-pmtiles/blob/main/rollup.config.js#L35-L44) to it can work with the web also. For [example](https://acalcutt.github.io/maplibre-contour-pmtiles/)

**Comment by HarelM** [comment]
It's probably worth mentioning that the feature added in #359 will not run in the worker thread.
There is still work to be done if one would like to support this. 
The best approach is to integrate it in maplibre as mentioned in this thread earlier...
